packages/core/components/Modal/ModalProvider.tsx

```diff
@@ -1,26 +1,91 @@
+import { nanoid } from 'nanoid';
 import * as React from 'react';
 
 import ModalContext, { Windows } from './ModalContext';
 
 export type ModalProviderProps = {};
 
-const ModalProvider: React.FunctionComponent<ModalProviderProps> = ({
-  children,
-}) => {
-  const [windows, setWindows] = React.useState<Array<Windows>>([]);
+type WindowStack = Record<string, Windows>;
+
+type WindowAction =
+  | {
+      type: 'ADD_WINDOW';
+      id: string;
+      window: Windows;
+    }
+  | {
+      type: 'UPDATE_WINDOW';
+      id: string;
+      window: Windows;
+    }
+  | {
+      type: 'REMOVE_WINDOW';
+      id: string;
+    }
+  | {
+      type: 'REMOVE_BY_TITLE';
+      title: string;
+    };
+
+const windowStackReducer: React.Reducer<WindowStack, WindowAction> = (
+  state,
+  action,
+) => {
+  const newWindows = { ...state };
+  switch (action.type) {
+    case 'ADD_WINDOW': {
+      newWindows[action.id] = action.window;
+      return newWindows;
+    }
+    case 'REMOVE_WINDOW': {
+      delete newWindows[action.id];
+      return newWindows;
+    }
+    case 'UPDATE_WINDOW': {
+      newWindows[action.id] = action.window;
+      return newWindows;
+    }
+    case 'REMOVE_BY_TITLE': {
+      Object.entries(newWindows).forEach(([id, window]) => {
+        if (window.title === action.title) {
+          delete newWindows[id];
+        }
+      });
+      return newWindows;
+    }
+    default:
+      return state;
+  }
+};
+
+const ModalProvider: React.FunctionComponent = ({ children }) => {
+  const [windows, dispatch] = React.useReducer(windowStackReducer, {});
   const [activeWindow, setActiveWindow] = React.useState<string>();
 
-  const addWindows = (window: Windows) =>
-    setWindows(state => [...state, window]);
-  const removeWindows = (title: string) =>
-    setWindows(state => state.filter(w => w.title !== title));
+  const addWindows = (window: Windows) => {
+    const id = nanoid();
+    dispatch({ type: 'ADD_WINDOW', id, window });
+    return id;
+  };
+  // TODO: Remove this method in the next major release - updating by IDs is more sensible
```

Comment:
Is this related to #194?

We could also remove it in the next minor release since your PR makes `removeWindows` obsolete.

Meanwhile, it is not a public API. ModalContext handles which window should be removed/updated so I don't think we should publish a major version about this. What do you think?

Reply:
While it wasn't intended to be part of a public API, it is accessible, so removing the method could break things for someone consuming using the provider directly for some purpose we haven't thought of. E.g. if a user wanted to manually manage what is shown in `Taskbar`

Reply:
I think a bit of additional work would be required to solve #194, since I tried to keep the rest of the API for this the same.

If we changed the API for the value of `ModalContext` to be the full object, with IDs, and used those IDs to manage the focus state, then I think that would solve #194

Reply:
Ah, I just re-read #194, yeah, this change would fix that problem. When I initially read it, I thought it was about focus states, which this doesn't deal with. The latest commit I've pushed deals with that issue, but if we want to keep the changes to the provider less breaking, I can pop that commit